File: node/src/PositionWatcher.js
const time = require('./Time')
const rpio = require('rpio')
const EventEmitter = require('events')
const Fiber = require('fibers')

module.exports = class PositionWatcher extends EventEmitter {

  // using https://www.npmjs.com/package/rpio
  constructor() {
    super()
    this.perimeter = 60*Math.PI
    this.wheelRatio = this.perimeter/2400

    this.theta = Math.PI/2
    this.x = 0
    this.y = 0
    
    this.ticks = [0, 0]
    this.oldTicks = [0, 0]
    
    this.sideState = [0, 0]
    this.sideOldState = [0, 0]

    this.backState = [0, 0]
    this.backOldState = [0, 0]

    this.isTurning = false

    this.L = 103.4
    this.l = 210
    
    this.backPhaseA = 6
    this.backPhaseB = 16

    this.sidePhaseC = 20
    this.sidePhaseD = 21

    this.enabled = false
  }

  openPins() {
    rpio.init({ mapping: 'gpio' })
    rpio.open(this.backPhaseA, rpio.INPUT)
    rpio.open(this.backPhaseB, rpio.INPUT)
    rpio.open(this.sidePhaseC, rpio.INPUT)
    rpio.open(this.sidePhaseD, rpio.INPUT)

    //rpio.read(15)
  }

  watchTicks() {
    return Fiber(() => {
      Fiber.yield(0)
      while (this.enabled) {
        let changed = false
        let sideFetchedState = [
          rpio.read(this.sidePhaseC),
          rpio.read(this.sidePhaseD)
        ]
        let backFetchedState = [
          rpio.read(this.backPhaseA),
          rpio.read(this.backPhaseB)
        ]
        //console.log(sideFetchedState, backFetchedState)

        if (sideFetchedState !== this.sideState) {
          this.sideState = sideFetchedState

          if (this.sideState[0] == this.sideOldState[1]) {
            this.ticks[0]--
          } else {
            this.ticks[0]++
          }
          
          this.sideOldState = this.sideState
          changed = true
        }

        if (backFetchedState !== this.backState) {
          this.backState = backFetchedState

          if (this.backState[0] == this.backOldState[1]) {
            this.ticks[1]--
          } else {
            this.ticks[1]++
          }

          this.backOldState = this.backState
          changed = true
        }

        if (changed) {
          Fiber.yield(this.ticks)
        }
      }
    }).run.bind(this)
  }

  async watchPosititon(fetchTicks) {
    let oldTicks = [0, 0]
    while (this.enabled) {
      // LEFT = SIDE
      // RIGHT = BACK
      let ticks = fetchTicks()
      if (ticks !== oldTicks) {
        let deltaTicks = [
          ticks[0] - oldTicks[0],
          ticks[1] - oldTicks[1]
        ]
        console.log(deltaTicks)
        oldTicks = ticks
        // let sideDistance = deltaTicks[0] * this.wheelRatio
        // let backDistance = deltaTicks[1] * this.wheelRatio
        // // if (this.isTurning) {
        // //   this.theta += (sideDistance/this.l + backDistance/this.L)/2
        // // } else {
          
        // // }
        // this.x += Math.sin(this.theta)*backDistance + Math.cos(this.theta)*sideDistance
        // this.y += Math.cos(this.theta)*backDistance + Math.sin(this.theta)*sideDistance
        // // posititon has changed, handle it ?
        // this.emit('positionUpdated', this.x, this.y, this.theta)
      }
      // wait ?
      await time.wait(10)
    }
  }

  init() {
    this.openPins()
    this.enabled = true
    console.log('> PositionWatcher: starting...')
    let ticks = this.watchTicks()
    console.log('> tick watcher started')
    this.watchPosititon(ticks)
  }

  stop() {
    this.enabled = false
  }
} 

File: python/src/PositionWatcher.py
from gpiozero import DigitalInputDevice
from .ThreadHelper import Thread
import time
from math import *

'''
This class manage all the odometry operations
'''
class PositionWatcher:
  backPerimeter = 88*pi
  lateralPerimeter = 60*pi
  
  # distance entre les deux encodeurs latéraux (milieux) (arrête de la base)
  axialDistance = 284 #280
  
  # distance entre l'encodeur arrirère et la droite qui passe par les deux encodeurs latéraux
  backAxialDistance = 110
  
  # coté bleu x: 979, y: 159
  # defaultX = 979
  # defaultY = 159
  defaultX = 979
  defaultY = 159
  defaultTheta = pi

  # left (scotch bleu) encodeur branché sur la prise du milieur
  phaseA = DigitalInputDevice(20, True) # 20 
  phaseB = DigitalInputDevice(21, True) # 21
  
  # right (sans scotch) encodeur branché côté carte sd
  phaseC = DigitalInputDevice(16, True) # 16
  phaseD = DigitalInputDevice(6, True) # 6

  # back  (scotch vert) encodeur branché coté port USB
  phaseE = DigitalInputDevice(5, True)
  phaseF = DigitalInputDevice(19, True)
  
  watchPositionThread = None
  watchTicksThread = None
  
  watchTicksEnabled = False
  watchPositionEnabled = False

  positionChangedHandler = None
  
  ignoreXChanges = False
  
  ignoreYChanges = False

  def __init__(self, container):
    self.logger = container.get('logger').get('PositionWatcher')
    self.reset(False)
    
  def setIgnoreXChanges(self, val):
    print('ignoreXChanges is now at', val)
    self.logger.info('ignoreXChanges is now at', val)
    self.ignoreXChanges = val
  
  def setIgnoreYChanges(self, val):
    print('ignoreYChanges is now at', val)
    self.logger.info('ignoreYChanges is now at', val)
    self.ignoreYChanges = val
  
  '''
  This thread will keep updated the left, right and back tick count
  '''
  def watchTicks(self):
    self.logger.info('WatchTicks thread START!')
    while self.watchTicksEnabled:
      leftFetchedState = (self.phaseA.value, self.phaseB.value)
      rightFetchedState = (self.phaseC.value, self.phaseD.value)
      backFetchedState = (self.phaseE.value, self.phaseF.value)
      
      if not self.ignoreXChanges:
        if leftFetchedState != self.leftState:
          self.leftState = leftFetchedState

          if self.leftState[0] == self.leftOldState[1]:
            self.leftTicks -= 1
          else:
            self.leftTicks += 1

          self.leftOldState = self.leftState

        if rightFetchedState != self.rightState:
          self.rightState = rightFetchedState

          if self.rightState[0] == self.rightOldState[1]:
            self.rightTicks -= 1
          else:
            self.rightTicks += 1

          self.rightOldState = self.rightState

      if not self.ignoreYChanges:
        if backFetchedState != self.backState:
          self.backState = backFetchedState

          if self.backState[0] == self.backOldState[1]:
            self.backTicks -= 1
          else:
            self.backTicks += 1

          self.backOldState = self.backState
    self.logger.info("WatchTicks thread QUIT!")

  '''
  /!\ Call once
  '''
  def computePosition(self):
    newTicks = (self.leftTicks, self.rightTicks, self.backTicks)
    if (newTicks != self.oldTicks):
      deltaTicks = (
        newTicks[0] - self.oldTicks[0],
        newTicks[1] - self.oldTicks[1],
        newTicks[2] - self.oldTicks[2]
      )
      self.oldTicks = newTicks
      
      leftDistance = deltaTicks[0] / 2400 * self.lateralPerimeter
      rightDistance = deltaTicks[1] / 2400 * self.lateralPerimeter
      backDistance = deltaTicks[2] / 2400 * self.backPerimeter

      #tb = (leftDistance + rightDistance) / 2
      #deltaTheta = 2 * asin((rightDistance - tb) / self.axialDistance)
      # print(self.axialDistance)
      deltaTheta = (rightDistance - leftDistance) / self.axialDistance
      
      backDistance -= deltaTheta*self.backAxialDistance
      rightDistance -= deltaTheta*self.axialDistance/2
      
      c = 0 if self.ignoreXChanges else 1
      
      self.theta += deltaTheta
      self.theta = self.theta % (2*pi)
      self.x += cos(self.theta) * rightDistance + sin(self.theta) * backDistance * c
      self.y += sin(self.theta) * rightDistance - cos(self.theta) * backDistance
      
      if self.positionChangedHandler != None:
        self.positionChangedHandler(self.x, self.y, self.theta)

    return (self.x, self.y, self.theta)

  def watchPosition(self):
    self.logger.info("WatchPosition thread START!")
    while self.watchPositionEnabled:
      self.computePosition()
      time.sleep(0.01)
    self.logger.info("WatchPosition thread QUIT!")

  def startWatchTicks(self):
    if not self.watchTicksEnabled:
      self.watchTicksEnabled = True
      self.watchTicksThread = Thread(target=self.watchTicks)
      self.watchTicksThread.start()
  
  def startWatchPosition(self):
    if not self.watchPositionEnabled:
      self.watchPositionEnabled = True
      self.watchPositionThread = Thread(target=self.watchPosition)
      self.watchPositionThread.start()
      
  def start(self):
    self.startWatchTicks()
    self.startWatchPosition()

  def stop(self):
    self.watchTicksEnabled = False
    self.watchPositionEnabled = False
    if self.watchTicksThread != None:
      self.watchTicksThread.stop()
    if self.watchPositionThread != None:
      self.watchPositionThread.stop()

  def pauseWatchPosition(self):
    self.watchPositionEnabled = False
    
  def resumeWatchPosition(self):
    #print('> PositionWatcher: resumed!')
    self.startWatchPosition()

  def setPositionChangedHandler(self, handler):
    self.positionChangedHandler = handler

  def getTicks(self):
    return (self.leftTicks, self.rightTicks, self.backTicks)
  
  def getPos(self):
    return (self.x, self.y, self.theta)

  def getData(self):
    return (self.x, self.y, self.theta)
  
  def setPos(self, x = None, y = None, theta = None):
    if x != None:
      self.x = x
    if y != None:
      self.y = y
    if theta != None:
      self.theta = theta
    self.logger.info('Now set to', {
      'x': round(self.x, 0),
      'y': round(self.y, 0),
      'theta': round(degrees(self.theta), 2)
    })

  def reset(self, log = True):
    self.x = self.defaultX
    self.y = self.defaultY
    self.theta = self.defaultTheta

    self.leftTicks = 0
    self.rightTicks = 0
    self.backTicks = 0

    self.leftState = (0, 0)
    self.leftOldState = (0, 0)

    self.rightState = (0, 0)
    self.rightOldState = (0, 0)
    
    self.backState = (0, 0)
    self.backOldState = (0, 0)

    self.oldTicks = (0, 0, 0)

    if log:
      self.logger.info("Reset done: position and orientation are at the default values")
